ddtrace/internal/datadog/profiling/dd_wrapper/include/libdatadog_helpers.hpp

```diff
@@ -0,0 +1,131 @@
+    X(thread_id, "thread id")                                                                                          \
+    X(thread_native_id, "thread native id")                                                                            \
+    X(thread_name, "thread name")                                                                                      \
+    X(task_id, "task id")                                                                                              \
+    X(task_name, "task name")                                                                                          \
+    X(span_id, "span id")                                                                                              \
+    X(local_root_span_id, "local root span id")                                                                        \
+    X(trace_type, "trace type")                                                                                        \
+    X(trace_resource_container, "trace resource container")                                                            \
+    X(trace_endpoint, "trace endpoint")                                                                                \
+    X(class_name, "class name")                                                                                        \
+    X(lock_name, "lock name")
+
+#define X_ENUM(a, b) a,
+#define X_STR(a, b) b,
+
+enum class ExportTagKey
+{
+    EXPORTER_TAGS(X_ENUM) _Length
+};
+
+enum class ExportLabelKey
+{
+    EXPORTER_LABELS(X_ENUM) _Length
+};
+
+// Utility functions
+inline ddog_CharSlice
+to_slice(std::string_view str)
+{
+    return { .ptr = str.data(), .len = str.size() };
+}
+
+inline std::string
+err_to_msg(const ddog_Error* err, std::string_view msg)
+{
+    auto ddog_err = ddog_Error_message(err);
+    std::string err_msg;
+    return std::string{ msg } + "(" + err_msg.assign(ddog_err.ptr, ddog_err.ptr + ddog_err.len) + ")";
+}
+
+inline static bool
+add_tag(ddog_Vec_Tag& tags, const ExportTagKey key, std::string_view val, std::string& errmsg)
+{
+    // NB the storage of `val` needs to be guaranteed until the tags are flushed
+    constexpr size_t num_keys = static_cast<size_t>(ExportTagKey::_Length);
+    constexpr std::array<std::string_view, num_keys> keys = { EXPORTER_TAGS(X_STR) };
+    std::string_view key_sv = keys[static_cast<size_t>(key)];
+
+    // Can't add empty tags. This isn't an error.
+    if (val.empty())
+        return true;
+
+    // Add
+    ddog_Vec_Tag_PushResult res = ddog_Vec_Tag_push(&tags, to_slice(key_sv), to_slice(val));
+    if (res.tag == DDOG_VEC_TAG_PUSH_RESULT_ERR) {
+        errmsg = err_to_msg(&res.err, "Error pushing tag");
+        errmsg += "(val:'" + std::string(val) + "')";
+        ddog_Error_drop(&res.err);
+        return false;
+    }
+    return true;
+}
+
+inline static bool
+add_tag_unsafe(ddog_Vec_Tag& tags, std::string_view key, std::string_view val, std::string& errmsg)
```

Comment:
I factored this differently as a result of other comments, but I changed the names without changing the fact that there are two species of `add_tag()`

variant 1:  takes an enum and a value.  This is safer because the enum is something we control and it's maintained in static storage.  The value may be passed by the user though, so I'd hardly call this "safe."  Mistake of naming in the original.

variant 2:  takes two string-like values.  Both are provided by the user, and each may have separate syntax checks in the backend.

I changed this so that 1 uses the inner implementation of 2 and just pulls the key names from static storage.